ogre2/src/Ogre2SegmentationCamera.cc

```diff
@@ -0,0 +1,648 @@
+
+            math::Color color;
+            if (label == this->backgroundLabel)
+              color = this->LabelToColor(label);
+            else
+              color = this->LabelToColor(compositeId);
+
+            subItem->setCustomParameter(1, Ogre::Vector4(
+              color.R(), color.G(), color.B(), 1.0));
+          }
+          else
+          {
+            // 256 => 8 bits .. 255 => color percentage
+            float labelColor = label / 255.0;
+            float instanceColor1 = (instanceCount / 256) / 255.0;
+            float instanceColor2 = (instanceCount % 256) / 255.0;
+
+            subItem->setCustomParameter(1, Ogre::Vector4(
+              labelColor, instanceColor1, instanceColor2, 1.0));
+          }
+        }
+
+        // check if it's an overlay material by assuming the
+        // depth check and depth write properties are off.
+        if (!datablock->getMacroblock()->mDepthWrite &&
+            !datablock->getMacroblock()->mDepthCheck)
+          subItem->setMaterial(this->plainOverlayMaterial);
+        else
+          subItem->setMaterial(this->plainMaterial);
+      }
+    }
+    itor.moveNext();
+  }
+
+  // reset the count & colors tracking
+  this->instancesCount.clear();
+  this->takenColors.clear();
+  this->coloredLabel.clear();
+}
+
+/////////////////////////////////////////////////
+void SegmentationMaterialSwitcher::postRenderTargetUpdate(
+    const Ogre::RenderTargetEvent &/*_evt*/)
+{
+  // restore item to use hlms material
+  auto itor = this->scene->OgreSceneManager()->getMovableObjectIterator(
+      Ogre::ItemFactory::FACTORY_TYPE_NAME);
+  while (itor.hasMoreElements())
+  {
+    Ogre::MovableObject *object = itor.peekNext();
+    Ogre::Item *item = static_cast<Ogre::Item *>(object);
+    for (unsigned int i = 0; i < item->getNumSubItems(); i++)
+    {
+      Ogre::SubItem *subItem = item->getSubItem(i);
+      auto it = this->datablockMap.find(subItem);
+      if (it != this->datablockMap.end())
+        subItem->setDatablock(it->second);
+    }
+    itor.moveNext();
+  }
+}
+
+/////////////////////////////////////////////////
+class ignition::rendering::Ogre2SegmentationCameraPrivate
+{
+  /// \brief Material Switcher to switch item's material
+  /// with colored version for segmentation
+  public: SegmentationMaterialSwitcher *materialSwitcher;
+
+  /// \brief Compositor Manager to create workspace
+  public: Ogre::CompositorManager2 *ogreCompositorManager;
+
+  /// \brief Workspace to interface with render texture
+  public: Ogre::CompositorWorkspace *ogreCompositorWorkspace;
+
+  /// \brief Workspace Definition
+  public: std::string workspaceDefinition;
+
+  /// \brief Render Texture to store the final segmentation data
+  public: Ogre::RenderTexture *ogreRenderTexture;
+
+  /// \brief Texture to create the render texture from.
+  public: Ogre::TexturePtr ogreTexture;
+
+  /// \brief Pixel Box to copy render texture data to a buffer
+  public: Ogre::PixelBox *pixelBox;
```

Comment:
Initialize all of these pointers to `nullptr`.  We've had bugs in the past when they weren't initialized.